test/decl/protocol/req/associated_type_inference_fixed_type.swift

```diff
@@ -435,19 +435,199 @@ do {
+// expected-error@-1 {{'Self.B' cannot be equal to both 'Never' and 'Int'}}
+// expected-note@-2 {{same-type constraint 'Self.B' == 'Int' implied here}}
+do {
+  // CHECK-LABEL: Abstract type witness system for conformance of Conformer to P31c: {
+  // CHECK-NEXT: A => (ambiguous), [[EQUIV_CLASS:0x[0-9a-f]+]]
+  // CHECK-NEXT: B => (ambiguous), [[EQUIV_CLASS]]
+  // CHECK-NEXT: }
+  struct Conformer: Q31 {}
+  // expected-error@-1 {{type 'Conformer' does not conform to protocol 'P31a'}}
+  // expected-error@-2 {{type 'Conformer' does not conform to protocol 'P31b'}}
+  // expected-error@-3 {{type 'Conformer' does not conform to protocol 'P31c'}}
+}
+
+protocol P32a where A == Int {
+  associatedtype A
+}
+protocol P32b where A == Bool {
+  associatedtype A
+}
+protocol P32c where B == Void {
+  associatedtype B
+}
+protocol P32d where B == Never {
+  associatedtype B
+}
+protocol P32e where A == B {
+  associatedtype A // expected-note {{protocol requires nested type 'A'; do you want to add it?}}
+  associatedtype B // expected-note {{protocol requires nested type 'B'; do you want to add it?}}
+}
+protocol Q32: P32e, P32a, P32b, P32c, P32d {}
+// expected-error@-1 {{'Self.B' cannot be equal to both 'Never' and 'Int'}}
+// expected-error@-2 {{'Self.B' cannot be equal to both 'Void' and 'Int'}}
+// expected-error@-3 {{'Self.A' cannot be equal to both 'Bool' and 'Int'}}
+// expected-note@-4 3 {{same-type constraint 'Self.A' == 'Int' implied here}}
+do {
+  // CHECK-LABEL: Abstract type witness system for conformance of Conformer to P32e: {
+  // CHECK-NEXT: A => (ambiguous), [[EQUIV_CLASS:0x[0-9a-f]+]]
+  // CHECK-NEXT: B => (ambiguous), [[EQUIV_CLASS]]
+  // CHECK-NEXT: }
+  struct Conformer: Q32 {}
+  // expected-error@-1 {{type 'Conformer' does not conform to protocol 'P32a'}}
+  // expected-error@-2 {{type 'Conformer' does not conform to protocol 'P32b'}}
+  // expected-error@-3 {{type 'Conformer' does not conform to protocol 'P32c'}}
+  // expected-error@-4 {{type 'Conformer' does not conform to protocol 'P32d'}}
+  // expected-error@-5 {{type 'Conformer' does not conform to protocol 'P32e'}}
+}
+
+protocol P33a where A == Int {
+  associatedtype A
+}
+protocol P33b where A == Int {
+  associatedtype A
+}
+protocol Q33: P33a, P33b {}
+do {
+  // CHECK-LABEL: Abstract type witness system for conformance of Conformer to P33a: {
+  // CHECK-NEXT: A => Int,
+  // CHECK-NEXT: }
+  struct Conformer: Q33 {}
+}
+
+protocol P34a {
+  associatedtype A = Void
+}
+protocol P34b {
+  associatedtype A = Never
+}
+protocol Q34a: P34a, P34b {}
+protocol Q34b: P34b, P34a {}
+protocol Q34c: P34a, P34b {
+  associatedtype A // expected-note {{protocol requires nested type 'A'; do you want to add it?}}
+}
+do {
+  // FIXME: should really be ambiguous (source-breaking)?
+  // CHECK-LABEL: Abstract type witness system for conformance of Conformer1 to P34a: {
+  // CHECK-NEXT: A => Void,
+  // CHECK-NEXT: }
+  struct Conformer1: Q34a {}
```

Comment:
It looks like this is sensitive to the order of protocols in the inheritance clause, which might be weird. Perhaps you should canonicalize the list of inherited protocols when walking (or walk over conformance requirements on 'Self' in the requirement signature instead, which are in canonical order).